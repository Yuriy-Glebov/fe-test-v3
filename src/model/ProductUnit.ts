
export const ProductUnit = {
    Kilo: 'Kilo',
    Each: 'Each'
}

export type ProductUnit = typeof ProductUnit[keyof typeof ProductUnit];
