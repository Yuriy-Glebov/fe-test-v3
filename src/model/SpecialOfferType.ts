
export const SpecialOfferType = {
    ThreeForTwo: 'ThreeForTwo',
    TenPercentDiscount: 'TenPercentDiscount',
    TwoForAmount: 'TwoForAmount',
    FiveForAmount: 'FiveForAmount'
}

export type SpecialOfferType = typeof SpecialOfferType[keyof typeof SpecialOfferType];


