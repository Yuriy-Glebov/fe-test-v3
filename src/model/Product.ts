import {ProductUnit} from "./ProductUnit"

export type Product = {
    name: string;
    unit: ProductUnit;
}

export function createProduct(name: string, unit: ProductUnit): Product {
    return {
      name,
      unit
    }
  }