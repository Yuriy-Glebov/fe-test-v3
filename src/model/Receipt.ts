import { Discount } from "./Discount";
import { Product } from "./Product";
import { createReceiptItem, ReceiptItem } from "./ReceiptItem";
import * as _ from "lodash";

export type Receipt = {
  getTotalPrice: () => number;
  addProduct: (
    p: Product,
    quantity: number,
    price: number,
    totalPrice: number
  ) => void;
  getItems: () => ReceiptItem[];
  addDiscount: (discount: Discount) => void;
  getDiscounts: () => Discount[];
};

export function createReceipt() {
  let items: ReceiptItem[] = [];
  let discounts: Discount[] = [];

  return {
    getTotalPrice: function () {
      let total = 0.0;
      for (let item of items) {
        total += item.totalPrice;
      }
      for (let discount of discounts) {
        total -= discount.discountAmount;
      }
      return total;
    },
    addProduct: function (
      p: Product,
      quantity: number,
      price: number,
      totalPrice: number
    ) {
      items.push(createReceiptItem(p, quantity, price, totalPrice));
    },
    getItems: function () {
      return _.clone(items);
    },
    addDiscount: function (discount: Discount) {
      discounts.push(discount);
    },
    getDiscounts: function () {
      return discounts;
    },
  };
}
