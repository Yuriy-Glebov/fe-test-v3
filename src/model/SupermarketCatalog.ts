import {Product} from "./Product"

export class SupermarketCatalog {
    private products: {[key: string]: Product} = {};
    private prices: {[key: string]: number} = {};

    public addProduct(product: Product, price: number): void {
        this.products[product.name] = product;
        this.prices[product.name] = price;
    }

    public getUnitPrice(p: Product): number {
        return this.prices[p.name];
    }
}
