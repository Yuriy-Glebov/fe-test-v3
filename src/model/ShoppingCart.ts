import { Product } from "./Product";
import { SupermarketCatalog } from "./SupermarketCatalog";
import * as _ from "lodash";
import { createProductQuantity, ProductQuantity } from "./ProductQuantity";
import { createDiscount, Discount } from "./Discount";
import { createReceipt, Receipt } from "./Receipt";
import { Offer } from "./Offer";
import { SpecialOfferType } from "./SpecialOfferType";

type ProductQuantities = { [productName: string]: ProductQuantity };
export type OffersByProduct = { [productName: string]: Offer };

export type ShoppingCart = {
  getItems: () => ProductQuantity[];
  addItem: (product: Product) => void;
  addItemQuantity: (product: Product, quantity: number) => void;
  handleOffers: (
    receipt: Receipt,
    offers: OffersByProduct,
    catalog: SupermarketCatalog
  ) => void;
};

export function createShoppingCart() {
  const items: ProductQuantity[] = [];
  const productQuantities: ProductQuantities = {};

  const addItemQuantity = (product: Product, quantity: number): void => {
    let productQuantity = createProductQuantity(product, quantity);
    items.push(productQuantity);
    let currentQuantity = productQuantities[product.name];
    if (currentQuantity) {
      productQuantities[product.name] = increaseQuantity(
        product,
        currentQuantity,
        quantity
      );
    } else {
      productQuantities[product.name] = productQuantity;
    }
  };

  const increaseQuantity = (
    product: Product,
    productQuantity: ProductQuantity,
    quantity: number
  ) => createProductQuantity(product, productQuantity.quantity + quantity);

  return {
    getItems: (): ProductQuantity[] => {
      return _.clone(items);
    },

    addItem: (product: Product): void => {
      addItemQuantity(product, 1.0);
    },
    addItemQuantity,

    handleOffers: (
      receipt: Receipt,
      offers: OffersByProduct,
      catalog: SupermarketCatalog
    ): void => {
      for (const productName in productQuantities) {
        const productQuantity = productQuantities[productName];
        const product = productQuantity.product;
        const quantity: number = productQuantities[productName].quantity;
        if (offers[productName]) {
          const offer: Offer = offers[productName];
          const unitPrice: number = catalog.getUnitPrice(product);
          let quantityAsInt = quantity;
          let discount: Discount | null = null;
          const x = 1;
          if (offer.offerType == SpecialOfferType.ThreeForTwo) {
            x = 3;
          } else if (offer.offerType == SpecialOfferType.TwoForAmount) {
            x = 2;
            if (quantityAsInt >= 2) {
              const total =
                offer.argument * Math.floor(quantityAsInt / x) +
                (quantityAsInt % 2) * unitPrice;
              const discountN = unitPrice * quantity - total;
              discount = createDiscount(
                product,
                "2 for " + offer.argument,
                discountN
              );
            }
          }
          if (offer.offerType == SpecialOfferType.FiveForAmount) {
            x = 5;
          }
          const numberOfXs = Math.floor(quantityAsInt / x);
          if (
            offer.offerType == SpecialOfferType.ThreeForTwo &&
            quantityAsInt > 2
          ) {
            const discountAmount =
              quantity * unitPrice -
              (numberOfXs * 2 * unitPrice + (quantityAsInt % 3) * unitPrice);
            discount = createDiscount(product, "3 for 2", discountAmount);
          }
          if (offer.offerType == SpecialOfferType.TenPercentDiscount) {
            discount = createDiscount(
              product,
              offer.argument + "% off",
              (quantity * unitPrice * offer.argument) / 100.0
            );
          }
          if (
            offer.offerType == SpecialOfferType.FiveForAmount &&
            quantityAsInt >= 5
          ) {
            const discountTotal =
              unitPrice * quantity -
              (offer.argument * numberOfXs + (quantityAsInt % 5) * unitPrice);
            discount = createDiscount(
              product,
              x + " for " + offer.argument,
              discountTotal
            );
          }
          if (discount != null) receipt.addDiscount(discount);
        }
      }
    },
  };
}
