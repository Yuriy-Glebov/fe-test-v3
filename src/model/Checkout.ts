import { SupermarketCatalog } from "./SupermarketCatalog";
import { OffersByProduct, ShoppingCart } from "./ShoppingCart";
import { Product } from "./Product";
import { createReceipt, Receipt } from "./Receipt";
import { createOffer } from "./Offer";
import { SpecialOfferType } from "./SpecialOfferType";

export type Checkout = {
  addSpecialOffer: (
    offerType: SpecialOfferType,
    product: Product,
    argument: number
  ) => void;
  checksOutArticlesFrom: (theCart: ShoppingCart) => Receipt;
};

export function createCheckout(catalog: SupermarketCatalog): Checkout {
  const offers: OffersByProduct = {};

  return {
    addSpecialOffer: (
      offerType: SpecialOfferType,
      product: Product,
      argument: number
    ) => {
      offers[product.name] = createOffer(offerType, product, argument);
    },
    checksOutArticlesFrom: (theCart: ShoppingCart): Receipt => {
      const receipt = createReceipt();
      const productQuantities = theCart.getItems();
      for (let pq of productQuantities) {
        let p = pq.product;
        let quantity = pq.quantity;
        let unitPrice = catalog.getUnitPrice(p);
        let price = quantity + unitPrice;
        receipt.addProduct(p, quantity, unitPrice, price);
      }
      theCart.handleOffers(receipt, offers, catalog);

      return receipt;
    },
  };
}
