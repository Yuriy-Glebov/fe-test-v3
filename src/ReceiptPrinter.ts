import { ProductUnit } from "./model/ProductUnit";
import { ReceiptItem } from "./model/ReceiptItem";
import { createReceipt, Receipt } from "./model/Receipt";

export function createReceiptPrinter(columns: number = 40) {
  return {
    printReceipt: (receipt: Receipt): string => {
      let result = "";
      for (const item of receipt.getItems()) {
        let price = format2Decimals(item.totalPrice);
        let quantity = presentQuantity(item);
        let name = item.product.name;
        let unitPrice = format2Decimals(item.price);

        let whitespaceSize = columns - name.length - price.length;
        let line =
          name + getWhitespace(whitespaceSize) + price + "\n";

        if (item.quantity != 1) {
          line += "  " + unitPrice + " * " + quantity + "\n";
        }
        result += line;
      }
      for (const discount of receipt.getDiscounts()) {
        let productPresentation = discount.product.name;
        let pricePresentation = format2Decimals(discount.discountAmount);
        let description = discount.description;
        result += description;
        result += "(";
        result += productPresentation;
        result += ")";
        result += getWhitespace(
          columns -
            3 -
            productPresentation.length -
            description.length -
            pricePresentation.length
        );
        result += "-";
        result += pricePresentation;
        result += "\n";
      }
      result += "\n";
      let pricePresentation = format2Decimals(receipt.getTotalPrice());
      let total = "Total: ";
      let whitespace = getWhitespace(
        columns - total.length - pricePresentation.length
      );
      result += total;
      result += whitespace;
      result += pricePresentation;

      return result;
    },
  };
}

function format2Decimals(number: number) {
  return new Intl.NumberFormat("en-UK", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(number);
}

function presentQuantity(item: ReceiptItem): string {
  return ProductUnit.Each == item.product.unit
    ? // TODO make sure this is the simplest way to make something similar to the java version
      new Intl.NumberFormat("en-UK", { maximumFractionDigits: 0 }).format(
        item.quantity
      )
    : new Intl.NumberFormat("en-UK", { minimumFractionDigits: 3 }).format(
        item.quantity
      );
}

function getWhitespace(whitespaceSize: number): string {
  return " ".repeat(whitespaceSize);
}
