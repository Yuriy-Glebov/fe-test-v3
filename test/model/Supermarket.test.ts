import { Product } from "../../src/model/Product";
import { SupermarketCatalog } from "../../src/model/SupermarketCatalog";
import { Receipt } from "../../src/model/Receipt";
import { ShoppingCart } from "../../src/model/ShoppingCart";
import { SpecialOfferType } from "../../src/model/SpecialOfferType";
import { ProductUnit } from "../../src/model/ProductUnit";
import { ReceiptPrinter } from "../../src/ReceiptPrinter";
import { printStandardReceipt } from "../helper/printStandartReceipt";
import { FakeCatalog } from "./FakeCatalog";
import { Checkout } from "../../src/model/Checkout";

describe("Supermarket", function () {
  let catalog: SupermarketCatalog;
  let cashier: Checkout;
  let theCart: ShoppingCart;
  let toothbrush: Product;
  let rice: Product;
  let apples: Product;
  let bananas: Product;
  let cherryTomatoes: Product;

  beforeEach(() => {
    catalog = new FakeCatalog();
    cashier = new Checkout(catalog);
    theCart = new ShoppingCart();

    toothbrush = new Product("toothbrush", ProductUnit.Each);
    catalog.addProduct(toothbrush, 0.99);
    rice = new Product("rice", ProductUnit.Each);
    catalog.addProduct(rice, 2.99);
    apples = new Product("apples", ProductUnit.Kilo);
    catalog.addProduct(apples, 1.99);
    bananas = new Product("bananas", ProductUnit.Each);
    catalog.addProduct(bananas, 0.25);
    cherryTomatoes = new Product("cherry tomato box", ProductUnit.Each);
    catalog.addProduct(cherryTomatoes, 0.69);
  });

  it("an_empty_shopping_cart_should_cost_nothing", function (this: any) {
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("one_normal_item", function (this: any) {
    theCart.addItem(toothbrush);
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("two_normal_items", function (this: any) {
    theCart.addItem(toothbrush);
    theCart.addItem(rice);
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("buy_two_get_one_free", function (this: any) {
    theCart.addItem(toothbrush);
    theCart.addItem(toothbrush);
    theCart.addItem(toothbrush);
    cashier.addSpecialOffer(
      SpecialOfferType.ThreeForTwo,
      toothbrush,
      catalog.getUnitPrice(toothbrush)
    );
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(receipt.getDiscounts().length).toEqual(1);

    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("buy_five_get_one_free", function (this: any) {
    theCart.addItem(toothbrush);
    theCart.addItem(toothbrush);
    theCart.addItem(toothbrush);
    theCart.addItem(toothbrush);
    theCart.addItem(toothbrush);
    cashier.addSpecialOffer(
      SpecialOfferType.ThreeForTwo,
      toothbrush,
      catalog.getUnitPrice(toothbrush)
    );
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("loose_weight_product", function (this: any) {
    theCart.addItemQuantity(apples, 0.5);
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("percent_discount", function (this: any) {
    theCart.addItem(rice);
    cashier.addSpecialOffer(SpecialOfferType.TenPercentDiscount, rice, 10.0);
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(receipt.getDiscounts().length).toEqual(1);

    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("xForY_discount", function (this: any) {
    theCart.addItem(cherryTomatoes);
    theCart.addItem(cherryTomatoes);
    cashier.addSpecialOffer(
      SpecialOfferType.TwoForAmount,
      cherryTomatoes,
      0.99
    );
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(receipt.getDiscounts().length).toEqual(1);

    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("FiveForY_discount", function (this: any) {
    theCart.addItemQuantity(apples, 5);
    cashier.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99);
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("FiveForY_discount_withSix", function (this: any) {
    theCart.addItemQuantity(apples, 6);
    cashier.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99);
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("FiveForY_discount_withSixteen", function (this: any) {
    theCart.addItemQuantity(apples, 16);
    cashier.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99);
    const receipt = cashier.checksOutArticlesFrom(theCart);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("FiveForY_discount_withFour", function (this: any) {
    theCart.addItemQuantity(apples, 4);
    cashier.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99);
    const receipt = cashier.checksOutArticlesFrom(theCart);

    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });
});
