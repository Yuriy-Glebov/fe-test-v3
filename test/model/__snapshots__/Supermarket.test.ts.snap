// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Supermarket an_empty_shopping_cart_should_cost_nothing 1`] = `
"
Total:                              0.00"
`;
